zcash_primitives/src/transaction/builder.rs

```diff
@@ -450,16 +451,16 @@ impl<R: RngCore + CryptoRng> Builder<R> {
                         (diversifier, g_d)
                     };
 
-                    let pk_d = {
+                    let (pk_d, payment_address) = loop {
                         let dummy_ivk = Fs::random(&mut self.rng);
-                        g_d.mul(dummy_ivk, &JUBJUB)
+                        let pk_d = g_d.mul(dummy_ivk, &JUBJUB);
+                        if let Some(addr) = PaymentAddress::from_parts(diversifier, pk_d.clone()) {
+                            break (pk_d, addr);
+                        }
```

Comment:
The spec indeed says that you should use a valid diversifier here (section 4.7.2 which refers to 4.2.2), so looping to find one is correct.